tests/libvmi/vmi.go

```diff
@@ -0,0 +1,132 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2020 Red Hat, Inc.
+ *
+ */
+
+package libvmi
+
+import (
+	k8sv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	k8smetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/rand"
+
+	kvirtv1 "kubevirt.io/client-go/api/v1"
+)
+
+type Option func(vmiOptions)
+
+const DefaultVmiName = "testvmi"
+
+type action func(vmi *kvirtv1.VirtualMachineInstance)
+type vmiOptions map[int]action
+
+const defaultTestGracePeriod int64 = 0
+
+const (
+	// Options Identifiers
+	optNone = iota
+
+	optContainerImageName
+	optEnableRng
+	optResourceMemory
+	optInterface
+	optNetwork
+	optTerminationGracePeriod
+
+	optCloudInitNoCloudUserData
+	optCloudInitNoCloudUserDataB64
+	optCloudInitNoCloudNetworkData
+	optCloudInitNoCloudNetworkDataB64
+)
+
+var defaultVmiOptions = vmiOptions{
+	optInterface: func(vmi *kvirtv1.VirtualMachineInstance) {
+		vmi.Spec.Domain.Devices.Interfaces = append(
+			vmi.Spec.Domain.Devices.Interfaces, primaryInterfaceDevice(),
+		)
+	},
+	optNetwork: func(vmi *kvirtv1.VirtualMachineInstance) {
+		vmi.Spec.Networks = append(vmi.Spec.Networks, *kvirtv1.DefaultPodNetwork())
+	},
+	optTerminationGracePeriod: func(vmi *kvirtv1.VirtualMachineInstance) {
+		t := defaultTestGracePeriod
+		vmi.Spec.TerminationGracePeriodSeconds = &t
+	},
+	optResourceMemory: func(vmi *kvirtv1.VirtualMachineInstance) {
+		vmi.Spec.Domain.Resources.Requests = k8sv1.ResourceList{
+			k8sv1.ResourceMemory: resource.MustParse("8192Ki"),
+		}
+	},
+}
+
+// WithRng adds `rng` to the the vmi devices.
+func WithRng() Option {
+	return func(o vmiOptions) {
+		o[optEnableRng] = func(vmi *kvirtv1.VirtualMachineInstance) {
+			vmi.Spec.Domain.Devices.Rng = &kvirtv1.Rng{}
+		}
+	}
+}
+
+// WithResourceMemory specifies the vmi memory resource.
+func WithResourceMemory(value string) Option {
+	return func(o vmiOptions) {
+		o[optResourceMemory] = func(vmi *kvirtv1.VirtualMachineInstance) {
+			vmi.Spec.Domain.Resources.Requests = k8sv1.ResourceList{
+				k8sv1.ResourceMemory: resource.MustParse(value),
+			}
+		}
+	}
+}
+
+// New instantiates a new VMI configuration,
+// building its properties based on the specified With* options.
+func New(namespace, name string, opts ...Option) *kvirtv1.VirtualMachineInstance {
+	options := defaultVmiOptions
+	for _, opt := range opts {
+		opt(options)
+	}
+
+	name += rand.String(48)
+	vmi := baseVmi(namespace, name)
+
+	for _, f := range options {
+		f(vmi)
+	}
+
+	return vmi
+}
+
+func baseVmi(namespace, name string) *kvirtv1.VirtualMachineInstance {
+	vmi := kvirtv1.NewVMIReferenceFromNameWithNS(namespace, name)
+	vmi.Spec = kvirtv1.VirtualMachineInstanceSpec{Domain: kvirtv1.DomainSpec{}}
+	vmi.TypeMeta = k8smetav1.TypeMeta{
+		APIVersion: kvirtv1.GroupVersion.String(),
+		Kind:       "VirtualMachineInstance",
+	}
+	return vmi
+}
+
+func primaryInterfaceDevice() kvirtv1.Interface {
```

Comment:
There is storage.go, why not network.go?

Reply:
There is not enough reasoning to do it yet. We will create it once another network related options will present itself.

Reply:
One would argue that that is how utils junkyard starts. We gotta keep an eye on those new methods appearing here.

Reply:
This is true for any code, bad or good. There is always an option to leave code worse or better, no matter what language, algorithm or structure it has.
With every addition you should question if cleaning and refactoring is needed, so I see not problem with moving this in the future when another network related option is introduced. I am not in favor of opening a file just for two defaults.

Reply:
It can go either way really; let's first understand if this is the direction we want to go.